File: test/api/python/ouroborous.py
###############################################################################
# Top contributors (to current version):
#   Daniel Larraz
#
# This file is part of the cvc5 project.
#
# Copyright (c) 2009-2025 by the authors listed in the file AUTHORS
# in the top-level source directory and their institutional affiliations.
# All rights reserved.  See the file COPYING in the top-level source
# directory for licensing information.
# #############################################################################
#
# "Ouroborous" test: does cvc5 read its own output?
#
# The "Ouroborous" test, named after the serpent that swallows its
# own tail, ensures that cvc5 can parse some input, output it again
#  (in any of its languages) and then parse it again.  The result of
# the first parse must be equal to the result of the second parse;
# both strings and expressions are compared for equality.
#
# To add a new test, simply add a call to run_test_string() under
# run_test(), below.  If you don't specify an input language,
# LANG_SMTLIB_V2 is used.  If your example depends on symbolic constants,
# you'll need to declare them in the "declarations" global, just
# below, in SMT-LIBv2 form (but they're good for all languages).
##

import cvc5

def parse(instr: str, input_language: str, output_language: str) -> str:
    assert input_language == "smt2"
    assert output_language == "smt2"

    declarations = """(set-logic ALL)
(declare-sort U 0)
(declare-fun f (U) U)
(declare-fun x () U)
(declare-fun y () U)
(assert (= (f x) x))
(declare-fun a () (Array U (Array U U)))
"""

    tm = cvc5.TermManager()
    solver = cvc5.Solver(tm)

    solver.setOption("input-language", input_language)
    solver.setOption("output-language", output_language)

    sm = cvc5.SymbolManager(tm)
    parser = cvc5.InputParser(solver, sm)

    # Parse declarations to bind symbols
    parser.setStringInput(cvc5.InputLanguage.SMT_LIB_2_6, declarations, "internal-buffer")
    while True:
        cmd = parser.nextCommand()
        if cmd.isNull():
            break
        # Invoke the command, which may bind symbols
        cmd.invoke(solver, sm)

    assert parser.done() # parser should be done

    parser.setStringInput(cvc5.InputLanguage.SMT_LIB_2_6, instr, "internal-buffer")
    term = parser.nextTerm()
    s = str(term)
    assert parser.nextTerm().isNull()  # next expr should be null
    return s

def translate(instr: str, input_language: str, output_language: str) -> str:
    print("="*50)
    print(f"translating from {input_language} to {output_language} this string:")
    print(instr)
    outstr = parse(instr, input_language, output_language)
    print("got this:")
    print(outstr)
    print(f"reparsing as {output_language}")
    poutstr = parse(outstr, output_language, output_language)
    assert outstr == poutstr
    print(f"got same expressions {outstr} and {poutstr}")
    print("="*50)
    return outstr

def run_test_string(instr: str, instr_language: str):
    print(f"\nstarting with: {instr}\n   in language {instr_language}")
    smt2str = translate(instr, instr_language, "smt2")
    print(f"in SMT2      : {smt2str}")
    outstr = translate(smt2str, "smt2", "smt2")
    print(f"to SMT2 : {outstr}\n")
    assert outstr == smt2str

def run_test():
    run_test_string("(= (f (f y)) x)", "smt2")
    run_test_string("(= ((_ extract 2 1) (bvnot (bvadd #b000 #b011))) #b10)", "smt2")
    run_test_string("((_ extract 2 0) (bvnot (bvadd (bvmul #b001 #b011) #b011)))", "smt2")

run_test()

File: test/api/java/Ouroborous.java
/******************************************************************************
 * Top contributors (to current version):
 *   Daniel Larraz
 *
 * This file is part of the cvc5 project.
 *
 * Copyright (c) 2009-2025 by the authors listed in the file AUTHORS
 * in the top-level source directory and their institutional affiliations.
 * All rights reserved.  See the file COPYING in the top-level source
 * directory for licensing information.
 * ****************************************************************************
 *
 * "Ouroborous" test: does cvc5 read its own output?
 *
 * The "Ouroborous" test, named after the serpent that swallows its
 * own tail, ensures that cvc5 can parse some input, output it again
 * (in any of its languages) and then parse it again.  The result of
 * the first parse must be equal to the result of the second parse;
 * both strings and expressions are compared for equality.
 *
 * To add a new test, simply add a call to runTestString() under
 * runTest(), below.  If you don't specify an input language,
 * LANG_SMTLIB_V2 is used.  If your example depends on symbolic constants,
 * you'll need to declare them in the "declarations" global, just
 * below, in SMT-LIBv2 form (but they're good for all languages).
 */

import io.github.cvc5.*;
import io.github.cvc5.modes.*;
import java.io.*;

public class Ouroborous
{
  public static void main(String[] args)
  {
    try
    {
      System.exit(runTest());
    }
    catch (CVC5ApiException e)
    {
      System.err.println(e.getMessage());
    }
    catch (Exception e)
    {
      System.err.println("non-cvc5 exception thrown");
      e.printStackTrace();
    }
    System.exit(1);
  }

  private static String parse(String instr, String inputLanguage, String outputLanguage)
      throws CVC5ApiException
  {
    assert inputLanguage.equals("smt2");
    assert outputLanguage.equals("smt2");

    String declarations = "(set-logic ALL)\n"
        + "(declare-sort U 0)\n"
        + "(declare-fun f (U) U)\n"
        + "(declare-fun x () U)\n"
        + "(declare-fun y () U)\n"
        + "(assert (= (f x) x))\n"
        + "(declare-fun a () (Array U (Array U U)))\n";

    TermManager tm = new TermManager();
    Solver solver = new Solver(tm);

    solver.setOption("input-language", inputLanguage);
    solver.setOption("output-language", outputLanguage);
    SymbolManager symman = new SymbolManager(tm);
    InputParser parser = new InputParser(solver, symman);

    parser.setStringInput(InputLanguage.SMT_LIB_2_6, declarations, "internal-buffer");

    Command c;
    while (true)
    {
      c = parser.nextCommand();
      if (c == null || c.isNull())
      {
        break;
      }
      c.invoke(solver, symman);
    }
    assert parser.done();

    parser.setStringInput(InputLanguage.SMT_LIB_2_6, instr, "internal-buffer");

    Term e = parser.nextTerm();
    String s = e.toString();
    assert parser.nextTerm() == null; // should be no more terms
    return s;
  }

  private static String translate(String instr, String inputLanguage, String outputLanguage)
      throws CVC5ApiException
  {
    assert inputLanguage.equals("smt2");
    assert outputLanguage.equals("smt2");

    System.out.println("==============================================");
    System.out.println(
        "translating from " + inputLanguage + " to " + outputLanguage + " this string:");
    System.out.println(instr);

    String outstr = parse(instr, inputLanguage, outputLanguage);

    System.out.println("got this:");
    System.out.println(outstr);
    System.out.println("reparsing as " + outputLanguage);

    String poutstr = parse(outstr, outputLanguage, outputLanguage);
    assert outstr.equals(poutstr);

    System.out.println("got same expressions " + outstr + " and " + poutstr);
    System.out.println("==============================================");

    return outstr;
  }

  private static void runTestString(String instr, String instrLanguage) throws CVC5ApiException
  {
    System.out.println();
    System.out.println("starting with: " + instr);
    System.out.println("   in language " + instrLanguage);

    String smt2str = translate(instr, instrLanguage, "smt2");
    System.out.println("in SMT2      : " + smt2str);

    String outstr = translate(smt2str, "smt2", "smt2");
    System.out.println("to SMT2 : " + outstr);
    System.out.println();

    assert outstr.equals(smt2str);
  }

  private static int runTest() throws CVC5ApiException
  {
    runTestString("(= (f (f y)) x)", "smt2");
    runTestString("(= ((_ extract 2 1) (bvnot (bvadd #b000 #b011))) #b10)", "smt2");
    runTestString("((_ extract 2 0) (bvnot (bvadd (bvmul #b001 #b011) #b011)))", "smt2");
    return 0;
  }
}